docs/USAGE-GUIDE.md

```diff
@@ -607,66 +607,41 @@ In TLS1.3, session ticket messages are sent after the handshake as "post-handsha
+`s2n_client_hello_get_session_id()` will get the session ID sent by the client in the ClientHello message. Note that this value may not be the session ID eventually associated with this particular connection since the session ID can change when the server sends the Server Hello. The official session ID can be retrieved with `s2n_connection_get_session_id()`.
 
-The callback can return a negative value to make s2n-tls terminate the
-handshake early with a fatal handshake failure alert.
+Call `s2n_client_hello_get_extensions()` to retrieve the entire list of extensions sent in the Client Hello. Calling `s2n_client_hello_get_extension_by_id()` allows you to interrogate the `s2n_client_hello` struct for a specific extension.
 
-```c
-int s2n_client_hello_cb_done(struct s2n_connection *conn)
-```
-Marks the non-blocking callback as complete.
-Can be invoked from within the callback when operating in non-blocking mode
-to continue the handshake.
+### Client Hello Callback
 
-```c
-int s2n_client_server_name_used(struct s2n_connection *conn)
-```
-Indicates that connection properties were changed on the basis of server_name.
-Triggers a s2n-tls server to send the server_name extension. Must be called
-before s2n-tls finishes processing the ClientHello.
+Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A possible use-case for this is to modify the `s2n_connection` based on information in the Client Hello. This should be done carefully as modifying the connection mid-handshake can be dangerous. In particular, switching from a `s2n_config` that supports TLS1.3 to one that does not opens the server up to a possible version downgrade attack. Note that `s2n_connection_server_name_extension_used()` MUST be invoked before exiting the callback if any of the connection properties were changed on the basis of the Server Name extension. If desired, the callback can return a negative value to make s2n-tls terminate the handshake early with a fatal handshake failure alert.
+
+#### Callback Modes
+
+The callback can be invoked in two modes: **S2N_CLIENT_HELLO_CB_BLOCKING** and **S2N_CLIENT_HELLO_CB_NONBLOCKING**. Use `s2n_config_set_client_hello_cb_mode()` to set the desired mode.
+
+The default mode, "blocking mode", will not return from the `s2n_negotiate()` call until the Client Hello callback returns. In this mode the callback should contain quick work like logging or simple configuration changes.
+ 
+In contrast, "non-blocking mode" will return from the `s2n_negotiate()` call if the Client Hello callback isn't finished yet, allowing the application to do work asynchronously. This mode should be used if the callback needs to make a network call or offload an expensive calculation to another thread. Only when the callback signals its work is complete by calling `s2n_client_hello_cb_done()` will the handshake continue.
```

Comment:
Not quite. Technically the s2n_negotiate call still returns after the Client Hello callback finishes, it just doesn't continue the handshake until s2n_client_hello_cb_done is called. The callback still has to return immediately.